### PR TITLE
Drees 102 biology act extraction

### PR DESCRIPTION
**Pull request by strayMat**
Adding an extractor for biology acts in the DCIR (all laboratory tests).

It needed a slight modification of the dcir test-input parquet file to add the ER_BIO_F table. It should not affect a DCIR flat table without ER_BIO_F.

### CODE REVIEW

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/events/BiologyAct.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package fr.polytechnique.cmap.cnam.etl.events
+
+import java.sql.Timestamp
+import org.apache.spark.sql.Row
+
+trait BiologyAct extends AnyEvent with EventBuilder {
+
+  override val category: EventCategory[BiologyAct]
+
+  def fromRow(
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/events/BiologyAct.scala]: Do you really need this method?

File: src/test/scala/fr/polytechnique/cmap/cnam/etl/extractors/prestations/PractitionerClaimSpecialityExtractorSuite.scala [hunk 2]
```diff
@@ -60,7 +60,7 @@ class PractitionerClaimSpecialityExtractorSuite extends SharedContext {
     val result = NonMedicalPractitionerClaimExtractor.extract(sources, nonMedicalSpeCodes.toSet)
 
     // Then
-    assertDSs(result, expected)
+    assertDSs(result, expected, debug = true)
```
Comment (ysebiat) [inline src/test/scala/fr/polytechnique/cmap/cnam/etl/extractors/prestations/PractitionerClaimSpecialityExtractorSuite.scala]: Turn off the debug flag.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActExtractor.scala [hunk 3]
```diff
@@ -0,0 +1,62 @@
+package fr.polytechnique.cmap.cnam.etl.extractors.acts
+
+import fr.polytechnique.cmap.cnam.etl.events.{BiologyAct, BiologyDcirAct, DcirAct, EventBuilder}
+import fr.polytechnique.cmap.cnam.etl.extractors.dcir.DcirExtractor
+import fr.polytechnique.cmap.cnam.etl.sources.Sources
+import org.apache.spark.sql.{DataFrame, Row}
+
+import scala.util.Try
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActExtractor.scala]: Should be the highest on the import order.

File: src/test/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActsSuite.scala [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+import fr.polytechnique.cmap.cnam.SharedContext
+import fr.polytechnique.cmap.cnam.etl.events.{BiologyAct, BiologyDcirAct, DcirAct, Event}
+import fr.polytechnique.cmap.cnam.etl.sources.Sources
+import fr.polytechnique.cmap.cnam.util.functions._
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema
+import org.apache.spark.sql.types._
+import org.scalatest.Matchers.{an, convertToAnyShouldWrapper}
+import org.scalatest.TryValues._
+
+import scala.util.Success
```
Comment (ysebiat) [inline src/test/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActsSuite.scala]: Should be the highest on the import order.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/events/BiologyAct.scala (hunk 1 above)
Comment (strayMat) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/events/BiologyAct.scala]: You are right. I removed this class and included DcirBiologyAct object in `MedicalAct.scala`. It does not change anything to the test. I think that this BiologyAct class was the cause of the bad coverage because I forgot to implement a test for this class.

File: src/test/scala/fr/polytechnique/cmap/cnam/etl/extractors/prestations/PractitionerClaimSpecialityExtractorSuite.scala (hunk 2 above)
Comment (strayMat) [reply src/test/scala/fr/polytechnique/cmap/cnam/etl/extractors/prestations/PractitionerClaimSpecialityExtractorSuite.scala]: done

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActExtractor.scala (hunk 3 above)
Comment (strayMat) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActExtractor.scala]: fixed

File: src/test/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActsSuite.scala (hunk 4 above)
Comment (strayMat) [reply src/test/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActsSuite.scala]: fixed

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActExtractor.scala [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+import fr.polytechnique.cmap.cnam.etl.events.{BiologyDcirAct, DcirAct, EventBuilder, MedicalAct}
+import fr.polytechnique.cmap.cnam.etl.extractors.dcir.DcirExtractor
+import fr.polytechnique.cmap.cnam.etl.sources.Sources
+import org.apache.spark.sql.{DataFrame, Row}
+
+
+object DcirBiologyActExtractor extends DcirExtractor[MedicalAct] {
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActExtractor.scala]: This class has a lot of duplicate code for the MedicalActExtractor. Can't we merge them to a super class to avoid that duplication?

Comment (strayMat) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirBiologyActExtractor.scala]: Yes you are right. I changed the DcirMedicalActExtractor class into a DcirActExtractor trait and implemented two objects DcirMedicalActExtractor and DcirBiologyActExtractor. 
This seems to work welll.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirMedicalActExtractor.scala [hunk 6]
```diff
@@ -4,18 +4,18 @@ package fr.polytechnique.cmap.cnam.etl.extractors.acts
...
 import fr.polytechnique.cmap.cnam.etl.extractors.dcir.DcirExtractor
 import fr.polytechnique.cmap.cnam.etl.sources.Sources
 
-object DcirMedicalActExtractor extends DcirExtractor[MedicalAct] {
+trait DcirActExtractor extends DcirExtractor[MedicalAct] {
 
   private final val PrivateInstitutionCodes = Set(4D, 5D, 6D, 7D)
-  override val columnName: String = ColNames.CamCode
-  override val eventBuilder: EventBuilder = DcirAct
+  //override val columnName: String = ColNames.CamCode
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirMedicalActExtractor.scala]: Delete this two lines.

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirMedicalActExtractor.scala [hunk 7]
```diff
@@ -62,3 +62,16 @@ object DcirMedicalActExtractor extends DcirExtractor[MedicalAct] {
...
+object DcirMedicalActExtractor extends DcirActExtractor  {
+  override val columnName: String = ColNames.CamCode
+  override val eventBuilder: EventBuilder = DcirAct
+}
+
+
+object DcirBiologyActExtractor extends DcirActExtractor  {
+  override val columnName: String = ColNames.BioCode
+  override val eventBuilder: EventBuilder = BiologyDcirAct
+  override def code = (row: Row) => row.getAs[Double](columnName).toString
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirMedicalActExtractor.scala]: Why is this here? Isn't already implemented?

Comment (strayMat) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirMedicalActExtractor.scala]: code asks for a String but this column is a Double

File: src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirMedicalActExtractor.scala (hunk 6 above)
Comment (strayMat) [reply src/main/scala/fr/polytechnique/cmap/cnam/etl/extractors/acts/DcirMedicalActExtractor.scala]: done

**Comment by strayMat** [comment]
Rebase done, ready to merge ! :D

**Comment by ysebiat** [comment]
Go.
